### DISCUSSION TITLE
User Prisma in middleware: Error: PrismaClient is unable to be run in the browser.

### DISCUSSION THREAD

**Discussion by agangdi** [vercel/next.js#40416] [Help]
I think _middleware is server-side code, but when I user prisma in middleware, Nextjs throws Error: PrismaClient is unable to be run in the browser.

this is my middleware.ts:
```
import type { NextFetchEvent, NextRequest } from 'next/server'
import {NextResponse} from "next/server";
import {getToken} from "next-auth/jwt";
import UserLocaleModel from "../models/UserLocaleModel";

export async function middleware(req: NextRequest, ev: NextFetchEvent, res: NextResponse) {
    const session = await getToken({ req, secret: process.env.SECRET })
    console.log("next js is sooooo.... good", req.nextUrl.pathname, req.nextUrl.locale )
    if (typeof window === 'undefined') {
        await UserLocaleModel.createOrUpdate(session, req.nextUrl.locale)
    }
    return res;
}
```

and this is UserLocaleModel:

```
import prisma from '../prisma/lib/prisma';

export default class UserLocaleModel{

  public static async createOrUpdate(session: any, locale: string) {
    if (!session?.user?.id) {
      return ;
    }
    // @ts-ignore
    return await prisma.$transaction(async (prisma: PrismaClient) => {
      const model = await prisma.userLocale.findUnique({
        where: {
          userId: session.user.id
        }
      })
      if (model) {
        return await prisma.userLocale.update({
          where: {
            userId: session.user.id
          },
          data: {
            locale
          }
        })
      }else{
        return await prisma.userLocale.create({
          data: {
            userId: session.user.id,
            locale
          }
        })
      }
    })
  }

}

```

and the error:
```
error - (middleware)/node_modules/.prisma/client/index-browser.js (667:0) @ <unknown>
Error: PrismaClient is unable to be run in the browser.
```

**Comment by icyJoseph** [comment]
I am not sure how, but the problem here is that because the edge is a limited API, or rather, a browser like environment in the cloud, most libraries do not really work out of the box. For example if a library uses `http` from Node.js it won't work.

What is happening here? That Prisma needs to have an edge version that can be used in the middleware. And they do have one:

https://github.com/prisma/prisma/blob/6ee1a062549e41d5505909a2e959334d294402dc/packages/client/src/generation/utils/buildInjectableEdgeEnv.ts#L12

Now the problem is how do we tell Prisma that the code is running on the edge? Perhaps the checks they do conflict with Next.js' environment setup. I'll try to read about this, but if you can figure out about Prisma at the edge and what is it that they check to conclude that the code is running on the edge, then we can solve this. For example, some might check `typeof window === 'undefined'` to assert that we are in Node.js, other looks for `process` to discard the browser, some look for ENV Vars, etc, etc. 

## EDIT

I found this article: https://www.prisma.io/blog/database-access-on-the-edge-8F0t1s1BqOJE, with this https://github.com/ruheni/prisma-edge-functions/ repository that document how to do this 😄

**Reply by agangdi** [reply]
Thanks very much

**Reply by agangdi** [reply]
It's still not fixed my trouble, but I implement the feature in some other way. And maybe this problem should fix from Prisma's side.

**Reply by serg-and** [reply]
Would you mind sharing how you solved it?
